[BUG] Unable to deserialize SearchRequest with _source field as an array

### What is the bug?
The error `org.opensearch.client.json.UnexpectedJsonEventException: Unexpected JSON event 'START_ARRAY' instead of '[START_OBJECT, KEY_NAME, VALUE_STRING, VALUE_TRUE, VALUE_FALSE]` is raised when trying to deserialize following json payload to `SearchRequest`.
```
{
  "query": {
    "match_all": {}
  },
  "_source": ["_id"],
  "size": 1
}
```
Context: We are implementing a kind of proxy service which accepts any OpenSearch requests and overwrite some query logic before passing them to the OpenSearch cluster. Previously, we use `opensearch-rest-high-level-client` with `XContentParser` to parse incoming requests but now we would like to migrate to `opensearch-java` library.

Note that if `_source` is removed or change to string field, it can parse successfully but we cannot do this as a workaround as we expect to accept any form of valid requests from clients.

### How can one reproduce the bug?
Execute following code and then will get `org.opensearch.client.json.UnexpectedJsonEventException`
```java
import com.fasterxml.jackson.core.JsonFactory;
import org.opensearch.client.json.jackson.JacksonJsonpMapper;
import org.opensearch.client.json.jackson.JacksonJsonpParser;
import org.opensearch.client.opensearch.core.SearchRequest;

import java.io.IOException;

public class Test {
    public static void main(String[] args) throws IOException {
        String jsonPayload = "{\n" +
                "  \"query\": {\n" +
                "    \"match_all\": {}\n" +
                "  },\n" +
                "  \"_source\": [\"_id\"],\n" +
                "  \"size\": 1\n" +
                "}\n";

        SearchRequest searchRequest = SearchRequest._DESERIALIZER.deserialize(new JacksonJsonpParser(new JsonFactory().createParser(jsonPayload)), new JacksonJsonpMapper());
    }
}
```

### What is the expected behavior?
We expect `SearchRequest._DESERIALIZER` can parse the json payload successfully.

### What is your host/environment?
- library = opensearch-java 2.11.1
- SDK = Amazon Correto 11
- OS = MacOS Somona 14.5

@dblock I run following command and got the below error, not sure how to fix it? Do you a document how to run codegen properly?

```
$ ./gradlew :java-codegen:downloadLatestSpec :java-codegen:run

> Configure project :java-client
=======================================
  Runtime JDK Version   : 11
  Gradle JDK Version    : 11
=======================================

> Configure project :java-codegen
=======================================
  Runtime JDK Version   : 11
  Gradle JDK Version    : 11
=======================================

> Task :java-codegen:downloadLatestSpec
Download https://github.com/opensearch-project/opensearch-api-specification/releases/download/main-latest/opensearch-openapi.yaml

> Task :java-codegen:run
[CodeGenerator] INFO  - Specification Location: /Users/chanononman/opensearch-java/java-codegen/opensearch-openapi.yaml
[CodeGenerator] INFO  - Eclipse Configuration: /Users/chanononman/opensearch-java/buildSrc/formatterConfig-generated.xml
[CodeGenerator] INFO  - Output Directory: /Users/chanononman/opensearch-java/java-client/src/generated/java
[OpenApiSpecification] INFO  - Parsing spec: /Users/chanononman/opensearch-java/java-codegen/opensearch-openapi.yaml
[CodeGenerator] FATAL - Unexpected Error
java.lang.IllegalArgumentException: Multiple types not yet supported: [boolean, null, number, object, string]
        at org.opensearch.client.codegen.openapi.OpenApiSchema.lambda$new$0(OpenApiSchema.java:89) ~[main/:?]
        at java.base/java.util.Optional.flatMap(Optional.java:294) ~[?:?]
        at org.opensearch.client.codegen.openapi.OpenApiSchema.<init>(OpenApiSchema.java:82) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiElement.lambda$children$3(OpenApiElement.java:116) ~[main/:?]
        at org.opensearch.client.codegen.utils.Maps$EntryMapper.map(Maps.java:68) ~[main/:?]
        at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:178) ~[?:?]
        at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[?:?]
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133) ~[?:?]
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at org.opensearch.client.codegen.utils.Maps.transform(Maps.java:58) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiElement.children(OpenApiElement.java:113) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiElement.children(OpenApiElement.java:96) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiComponents.<init>(OpenApiComponents.java:30) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiElement.child(OpenApiElement.java:67) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiSpecification.<init>(OpenApiSpecification.java:68) ~[main/:?]
        at org.opensearch.client.codegen.openapi.OpenApiSpecification.retrieve(OpenApiSpecification.java:47) ~[main/:?]
        at org.opensearch.client.codegen.CodeGenerator.parseSpec(CodeGenerator.java:113) ~[main/:?]
        at org.opensearch.client.codegen.CodeGenerator.main(CodeGenerator.java:84) [main/:?]

> Task :java-codegen:run FAILED

FAILURE: Build failed with an exception.

Thank @Xtansia for your helps.
One thing I see when executing `:java-codegen:run` is it visits only `info` group. I'm not sure if it is supposed to generate code for every groups or not? And when I try changing `OPERATION_MATCHER` in [CodeGenerator](https://github.com/opensearch-project/opensearch-java/blob/0cf11db08a574963ab1f0c47bfc6ebb53fd73b9d/java-codegen/src/main/java/org/opensearch/client/codegen/CodeGenerator.java#L36) to **`search`**, I got this error.

Code:
```
private static final OperationGroup.Matcher OPERATION_MATCHER = OperationGroup.matcher().add(null, "search");
```

Error:
```
$ ./gradlew :java-codegen:run

> Configure project :java-client
=======================================
  Runtime JDK Version   : 11
  Gradle JDK Version    : 11
=======================================

> Configure project :java-codegen
=======================================
  Runtime JDK Version   : 11
  Gradle JDK Version    : 11
=======================================

> Task :java-codegen:run
[CodeGenerator] INFO  - Specification Location: /Users/chanononman/ips/opensearch-java/java-codegen/opensearch-openapi.yaml
[CodeGenerator] INFO  - Eclipse Configuration: /Users/chanononman/ips/opensearch-java/buildSrc/formatterConfig-generated.xml
[CodeGenerator] INFO  - Output Directory: /Users/chanononman/ips/opensearch-java/java-client/src/generated/java
[OpenApiSpecification] INFO  - Parsing spec: /Users/chanononman/ips/opensearch-java/java-codegen/opensearch-openapi.yaml
[SpecTransformer] INFO  - Visiting Specification: org.opensearch.client.codegen.openapi.OpenApiSpecification@6f1ff437[location=/Users/chanononman/ips/opensearch-java/java-codegen/opensearch-openapi.yaml]
[SpecTransformer] INFO  - Visiting Operation Group: search [4 variants]
[SpecTransformer] INFO  - Visiting Parameter: org.opensearch.client.codegen.openapi.OpenApiParameter@6f2c2e16[pointer=/components/parameters/search::path.index]
[SpecTransformer] INFO  - Visiting Parameter: org.opensearch.client.codegen.openapi.OpenApiParameter@4bb62c05[pointer=/components/parameters/search::query._source]
[CodeGenerator] FATAL - Unexpected Error
java.lang.UnsupportedOperationException: Can not get type name for oneOf: [org.opensearch.client.codegen.openapi.OpenApiSchema@45716ad8[pointer=/components/schemas/_core.search:SourceConfigParam/oneOf/0]]
        at org.opensearch.client.codegen.model.SpecTransformer.mapOneOf(SpecTransformer.java:375) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.mapTypeInner(SpecTransformer.java:329) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.mapType(SpecTransformer.java:304) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.mapType(SpecTransformer.java:298) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.mapTypeInner(SpecTransformer.java:315) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.mapType(SpecTransformer.java:304) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.mapType(SpecTransformer.java:298) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.visit(SpecTransformer.java:214) ~[main/:?]
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[?:?]
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274) ~[?:?]
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:502) ~[?:?]
        at org.opensearch.client.codegen.model.SpecTransformer.visit(SpecTransformer.java:188) ~[main/:?]
        at org.opensearch.client.codegen.model.SpecTransformer.visit(SpecTransformer.java:97) ~[main/:?]
        at java.base/java.util.HashMap.forEach(HashMap.java:1337) ~[?:?]
        at org.opensearch.client.codegen.model.SpecTransformer.visit(SpecTransformer.java:89) ~[main/:?]
        at org.opensearch.client.codegen.CodeGenerator.parseSpec(CodeGenerator.java:115) ~[main/:?]
        at org.opensearch.client.codegen.CodeGenerator.main(CodeGenerator.java:84) [main/:?]

> Task :java-codegen:run FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':java-codegen:run'.
> Process 'command '/Users/chanononman/.gradle/jdks/eclipse_adoptium-11-x86_64-os_x.2/jdk-11.0.24+8/Contents/Home/bin/java'' finished with non-zero exit value 1
```

--- a/.github/workflows/test-integration.yml
+++ b/.github/workflows/test-integration.yml
@@ -41,8 +41,8 @@ jobs:
       - name: Run Docker
         run: |
           echo "PASSWORD=admin" >> $GITHUB_ENV
-          docker-compose --project-directory .ci/opensearch build --build-arg OPENSEARCH_VERSION=${{ matrix.entry.opensearch_version }}
-          docker-compose --project-directory .ci/opensearch up -d
+          docker compose --project-directory .ci/opensearch build --build-arg OPENSEARCH_VERSION=${{ matrix.entry.opensearch_version }}
+          docker compose --project-directory .ci/opensearch up -d
           sleep 60
 
       - name: Sets password (new versions)
@@ -64,4 +64,4 @@ jobs:
 
       - name: Stop Docker
         run: |
-          docker-compose --project-directory .ci/opensearch down
+          docker compose --project-directory .ci/opensearch down

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -42,6 +42,7 @@ This section is for maintaining a changelog for all breaking changes for the cli
 
 ### Fixed
 - Fixed error when deserializing a normalizer without 'type' ([#1111](https://github.com/opensearch-project/opensearch-java/pull/1111))
+- Fixed deserialization of SearchRequest when `_source` is an array ([#1117](https://github.com/opensearch-project/opensearch-java/pull/1117))
 
 ### Security
 

--- a/java-client/src/main/java/org/opensearch/client/json/ObjectDeserializer.java
+++ b/java-client/src/main/java/org/opensearch/client/json/ObjectDeserializer.java
@@ -239,7 +239,12 @@ public class ObjectDeserializer<ObjectType> implements JsonpDeserializer<ObjectT
             throw new NoSuchElementException("No deserializer was setup for '" + name + "'");
         }
 
-        acceptedEvents = EventSetObjectAndString;
+        this.acceptedEvents = EventSetObjectAndString.clone();
+
+        if (this.shortcutProperty instanceof FieldObjectDeserializer) {
+            JsonpDeserializer<?> shortcutDeserializer = ((FieldObjectDeserializer<?, ?>) this.shortcutProperty).deserializer;
+            this.acceptedEvents.addAll(shortcutDeserializer.nativeEvents());
+        }
     }
 
     // ----- Object types

--- a/java-client/src/main/java/org/opensearch/client/json/UnionDeserializer.java
+++ b/java-client/src/main/java/org/opensearch/client/json/UnionDeserializer.java
@@ -179,8 +179,13 @@ public class UnionDeserializer<Union, Kind, Member> implements JsonpDeserializer
                 );
                 objectMembers.add(member);
                 if (od.shortcutProperty() != null) {
-                    // also add it as a string
-                    addMember(Event.VALUE_STRING, tag, member);
+                    // also add it as the shortcut property events
+                    for (Event e : od.acceptedEvents()) {
+                        if (e == Event.START_OBJECT || e == Event.KEY_NAME) {
+                            continue;
+                        }
+                        addMember(e, tag, member);
+                    }
                 }
             } else {
                 UnionDeserializer.SingleMemberHandler<Union, Kind, Member> member = new SingleMemberHandler<>(tag, deserializer);

--- a/java-client/src/test/java/org/opensearch/client/opensearch/core/SearchRequestTest.java
+++ b/java-client/src/test/java/org/opensearch/client/opensearch/core/SearchRequestTest.java
@@ -9,9 +9,12 @@
 package org.opensearch.client.opensearch.core;
 
 import java.util.Collections;
+import java.util.List;
 import org.junit.Test;
 import org.opensearch.client.json.JsonData;
 import org.opensearch.client.opensearch._types.FieldValue;
+import org.opensearch.client.opensearch.core.search.SourceConfig;
+import org.opensearch.client.opensearch.core.search.SourceFilter;
 import org.opensearch.client.opensearch.model.ModelTestCase;
 
 public class SearchRequestTest extends ModelTestCase {
@@ -56,5 +59,77 @@ public class SearchRequestTest extends ModelTestCase {
         SearchRequest copied = origin.toBuilder().build();
 
         assertEquals(copied.index(), origin.index());
+    }
+
+    @Test
+    public void canDeserializeSourceAsBoolean() {
+        String json = "{\"query\":{\"match_all\":{}},\"_source\":true,\"size\":1}";
+
+        SearchRequest searchRequest = fromJson(json, SearchRequest._DESERIALIZER);
+
+        SourceConfig _source = searchRequest.source();
+        assertNotNull(_source);
+        assertTrue(_source.isFetch());
+        assertTrue(_source.fetch());
+    }
+
+    @Test
+    public void canDeserializeSourceAsString() {
+        String json = "{\"query\":{\"match_all\":{}},\"_source\":\"_id\",\"size\":1}";
+
+        SearchRequest searchRequest = fromJson(json, SearchRequest._DESERIALIZER);
+
+        SourceConfig _source = searchRequest.source();
+        assertNotNull(_source);
+        assertTrue(_source.isFilter());
+
+        SourceFilter filter = _source.filter();
+        assertNotNull(filter);
+
+        List<String> includes = filter.includes();
+        assertEquals(1, includes.size());
+        assertEquals("_id", includes.get(0));
+
+        assertTrue(filter.excludes().isEmpty());
+    }
+
+    @Test
+    public void canDeserializeSourceAsArray() {
+        String json = "{\"query\":{\"match_all\":{}},\"_source\":[\"_id\"],\"size\":1}";
+
+        SearchRequest searchRequest = fromJson(json, SearchRequest._DESERIALIZER);
+
+        SourceConfig _source = searchRequest.source();
+        assertNotNull(_source);
+        assertTrue(_source.isFilter());
+
+        SourceFilter filter = _source.filter();
+        assertNotNull(filter);
+
+        List<String> includes = filter.includes();
+        assertEquals(1, includes.size());
+        assertEquals("_id", includes.get(0));
+
+        assertTrue(filter.excludes().isEmpty());
+    }
+
+    @Test
+    public void canDeserializeSourceAsObject() {
+        String json = "{\"query\":{\"match_all\":{}},\"_source\":{\"includes\":[\"_id\"]},\"size\":1}";
+
+        SearchRequest searchRequest = fromJson(json, SearchRequest._DESERIALIZER);
+
+        SourceConfig _source = searchRequest.source();
+        assertNotNull(_source);
+        assertTrue(_source.isFilter());
+
+        SourceFilter filter = _source.filter();
+        assertNotNull(filter);
+
+        List<String> includes = filter.includes();
+        assertEquals(1, includes.size());
+        assertEquals("_id", includes.get(0));
+
+        assertTrue(filter.excludes().isEmpty());
     }
 }

